agent/src/util/device_plugin_service.rs

```diff
@@ -476,27 +543,93 @@ async fn try_update_instance_device_usage(
         }
 
         // Update the instance to reserve this slot for this node iff it is available and not already reserved for this node.
-        if slot_available_to_reserve(device_usage_id, node_name, &instance)? {
-            instance
-                .device_usage
-                .insert(device_usage_id.to_string(), node_name.to_string());
+        let device_usage_string = instance.device_usage.get(device_usage_id);
+        if device_usage_string.is_none() {
+            // No corresponding id found
+            trace!(
+                "try_update_instance_device_usage - could not find {} id in device_usage",
+                device_usage_id
+            );
+            return Err(Status::new(
+                Code::Unknown,
+                "Could not find device usage slot",
+            ));
+        }
 
-            if let Err(e) = kube_interface
-                .update_instance(&instance, instance_name, instance_namespace)
-                .await
-            {
-                if x == (MAX_INSTANCE_UPDATE_TRIES - 1) {
-                    trace!("try_update_instance_device_usage - update_instance returned error [{}] after max tries ... returning error", e);
-                    return Err(Status::new(Code::Unknown, "Could not update Instance"));
+        let device_usage = DeviceUsage::from_str(device_usage_string.unwrap()).map_err(|_| {
+            Status::new(
+                Code::Unknown,
+                format!(
+                    "Fails to parse {} to DeviceUsage ",
+                    device_usage_string.unwrap()
+                ),
+            )
+        })?;
+        match get_device_usage_state(&device_usage, node_name) {
```

Comment:
if this match were in its own function it might make creating tests easier ... ignore if this match already has good test coverage :)

Reply:
yes, we have good test coverage for both try_update_instance_device_usage and get_device_usage_state